File: solutions/go/binary-search/1/binary_search.go
package binarysearch

func SearchInts(list []int, key int) int {
	mid := len(list) / 2
    if mid >= len(list) {
        return -1
    }
    midElem := list[mid]
    switch {
        case key < midElem:
        	return SearchInts(list[:mid], key)
        case key == midElem:
        	return mid
    }
    result := SearchInts(list[mid+1:], key)
    if (result != -1) {
        result += mid + 1
    }
    return result
}

File: solutions/c/binary-search/1/binary_search.c
#include "binary_search.h"

const int *binary_search(const int value, const int *arr, const size_t length) {
    if (!length) {
        return NULL;
    }
    const size_t mid = length / 2;
    const int mid_elem = arr[mid];
    if (value < mid_elem) {
        return binary_search(value, arr, mid);
    } else if (value == mid_elem) {
        return &arr[mid];
    } else {
        return binary_search(value, &arr[mid+1], length - mid - 1);
    }
}